Poor Man's Type Checking in Arguments and Returns

julitrows opened the pull request:
Please, disregard the branch name 😮‍💨 

### Changes

### Add optional type checking to `argument` declarations

```rb
argument :jarl, type: Chiquito
argument :torpedo, type: Array, default: nil
```

### Add optional type checking to `returned` values

```rb
returns MyClass, nullable: true
returns MyCollection, of: Item, nullable: true, allow_nils: false
```

## Notes
 
Made them **optional**, since they're not the main purpose of the library, but happy to cut a 3.0.0 branch with **breaking** changes.

See each commit's README entries for more examples

```rb
class UserByPermissionsQuery
  include Injectable
  
  argument(:permission, type: Permission)
  dependency(:base_relation) { User.all }
  returns(ActiveRecord::Associations::CollectionProxy, of: User, nullable: false, allow_nils: false)

  def call
    base_relation.includes(:permissions).where(permissions: { id: permission.id })
  end
end
```

### About Collection validations

I've got a bit of a mess there. Not really sure I should check for Enumerable, or just `responds_to?(:each)` is enough?

Maybe having an optional `enumerates_with:` key is cool to indicate which method is used to iterate the collection (default `:each`)?

```rb
returns MyCollection, of: Item, enumerates_with: :each, nullable: false, allow_nils: false
```

julitrows committed:
feat: add optional type field to argument declaration

julitrows committed:
feat: add optional typecheck to returned value

julitrows committed:
chore: bump version

julitrows commented on lib/injectable/validators/argument_declaration.rb:
```diff
@@ -0,0 +1,33 @@
+        def validate!(name, type = nil, default = nil)
+          return unless type
+
+          raise(ArgumentError, wrong_type_message(name)) unless type.is_a?(Module)
```
Confused: Should this be Module or Class or both?

julitrows commented on lib/injectable/class_methods.rb:
```diff
@@ -153,5 +180,26 @@ def required_call_arguments
+    def initialize_collection_return(collection_class, element_class)
+      raise(ArgumentError, ':of for returns must be a Class or Module') unless element_class.is_a?(Module)
+
+      valid_collection = collection_class.is_a?(Class) || collection_class.is_a?(Module)
```
Confusion: should I validate for Class and Module? Just for Class, just for Module?

julitrows commented on lib/injectable/class_methods.rb:
```diff
@@ -153,5 +180,26 @@ def required_call_arguments
+      valid_collection = collection_class.is_a?(Class) || collection_class.is_a?(Module)
+      raise(ArgumentError, ':collection for returns must be a Class or Module') unless valid_collection
+
+      valid_collection = collection_class.include?(Enumerable) # check method_defined?(:each) instead?
```
Here's the thing about Enumerable vs "ducktyping"

julitrows committed:
Update README.md

julitrows committed:
add line to changelog

julitrows committed:
refactor: check agasint class being a module, since Class < Module, for shortness

julitrows committed:
refactor: duck-type against :each being an instance method

julitrows commented:
@iovis WDYT?

aalbagarcia commented on README.md:
````diff
@@ -332,18 +332,187 @@ argument :browser, default: 'Unknown'
+```rb
+argument :user,   type: User
+argument :values, type: Array
+argument :report, type: Hash
````
I'd add different examples here, so a user looking at them can get a quick idea of what options you have:

```ruby
argument :user,   type: User
argument :values, type: Array, default: %i[done closed accepted rejected]
argument :report, type: Hash, default: nil
```

I know that the options are documented in the paragraph right below, but reading this code helps get a quick idea of which arguments are available and what I can do.

aalbagarcia commented on lib/injectable/class_methods.rb:
```diff
@@ -161,5 +184,26 @@ def required_call_arguments
+              "#{collection_class} is not a collection-like class (must include Enumerable) when specifying :of")
+      end
+
+      { kind: :collection, collection: collection_class, elem: element_class }
```
I'd named the keys  `:collection_type` and `:element_type`.

Later, when you use the `return_spec` value:

```ruby 
Injectable::Validators::CollectionReturns.validate!(
              *spec.values_at(:collection, :elem, :nullable, :allow_nils), result
            )
``` 

it looks like you are passing a collection and an element to `validate!`, this would be more clear:

```ruby 
Injectable::Validators::CollectionReturns.validate!(
              *spec.values_at(:collection_type, :element_type, :nullable, :allow_nils), result
            )
``` 

and it'd be aligned with the return spec of the method `initialize_single_return`, that uses the keys `:kind` and `:type`:

```ruby
def validate!(collection_type, element_type, nullable, allow_nils, result)

end
```

Finally, using `:collection_type` and `:element_type`, would  be consistent with the keys returned by  `initialize_single_return ` method, that uses `:kind`, and `:type`.

aalbagarcia commented on lib/injectable/validators/argument_declaration.rb:
```diff
@@ -0,0 +1,33 @@
+          ":type for argument #{name} must be a Class or Module"
+        end
+
+        def bad_default_type(name, default_class, type)
```
for consistency, I'd name this method `bad_default_type_message`.

aalbagarcia reviewed (dismissed):
Hey Julio!!

It took me some time but I finally did it.

To answer your questions. I prefer checking for `:each`. It's more *ruby-ish* than chekcing for `Enumerable`. So I'd keep it as you did in the lastest changes.

I'd keep the type checking optional. It would be a bit unnatural for me to use a library in ruby that forces me to use types. Also agree with iovis comment about the impact on performance, it would be something to double check if types become a non-optional thing.

Finally, I'm thinking of a use case that might be needed: what If I need the argument to be of different types?

```ruby
class MyClass
  include Injectable

  argument :user, type: [User, AdminUser]
end
```

I guess the answer would be to add two arguments and a validation to check that at least one exist:

```ruby
class MyClass
  include Injectable

  argument :user, type: User, default: nil
  argument :admin_user, type: AdminUser, default nil

  validate :user_exists

end
```

Maybe something to add to the readme if you don't want to implement this use case.

It looks good to me. Left a couple of comments about naming, but they are not blocking.

julitrows committed:
chore(readme): better examples

julitrows committed:
refactor: rename method for consistency

julitrows committed:
refactor: rename keys for consistency and clarity

julitrows committed:
chore(lint): autocorrect Style/ModuleMemberExistenceCheck

julitrows committed:
feature: argument declarations can also be arrays

argument :jarl, type: [String, Symbol]

call(jarl: 'fistro') # => ok
call(jarl: :fistro) # => ok

julitrows committed:
feat: output position of wrong element in a returned collection

specs pending

julitrows committed:
chore(lint): fix Style/PredicateWithKind

julitrows committed:
chore(lint): rubocop

julitrows committed:
chore(lint): rubocop RSpec/NestedGroups

julitrows committed:
chore(lint): rubocop
